pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/util/codearea/syntaxhighlighting/JavaSyntaxHighlighter.java

```diff
@@ -0,0 +1,79 @@
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.NULL;
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.NUMBER;
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.PAREN;
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.SEMICOLON;
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.SINGLEL_COMMENT;
+import static net.sourceforge.pmd.util.fxdesigner.util.codearea.syntaxhighlighting.HighlightClasses.STRING;
+
+import java.util.Arrays;
+import java.util.regex.Pattern;
+
+import net.sourceforge.pmd.util.fxdesigner.util.codearea.SimpleRegexSyntaxHighlighter;
+
+/**
+ * Syntax highlighter for Java.
+ *
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+public class JavaSyntaxHighlighter extends SimpleRegexSyntaxHighlighter {
+
+
+    private static final String[] KEYWORDS = new String[] {
+        "public", "return", "final", "import", "static", "new",
+        "extends", "int", "throws?", "void", "if", "this",
+        "private", "class", "else", "case", "package", "abstract",
+        "boolean", "break", "byte", "catch", "char", "for",
+        "continue", "default", "double", "enum", "finally",
+        "float", "implements", "instanceof", "interface", "long",
+        "native", "protected", "while", "assert", "short", "super",
+        "switch", "synchronized", "transient", "try", "volatile",
+        "do", "strictfp", "goto", "const", "open", 
+        "module", "requires", "transitive", "exports", 
+        "opens", "to", "uses", "provides", "with",
+        };
+
+
+    /** First characters of the keywords, used to optimise the regex. */
+    private static final String KEYWORDS_START_CHARS = Arrays.stream(KEYWORDS)
+                                                             .map(s -> s.substring(0, 1))
+                                                             .distinct()
+                                                             .reduce((s1, s2) -> s1 + s2)
+                                                             .get();
+
+
+    private static final RegexHighlightGrammar GRAMMAR
+        = grammarBuilder(SINGLEL_COMMENT.css, "//[^\n]*")
+        .or(MULTIL_COMMENT.css, "/\\*.*?\\*/")
+        .or(PAREN.css, "[()]")
+        .or(NUMBER.css, "\\b\\d+[fdlFDL]*\\b")
+        .or(BRACE.css, "[{}]")
+        .or(BRACKET.css, "[\\[]]")
+        .or(SEMICOLON.css, ";")
+        .or(KEYWORD.css, "\\b(?=[" + KEYWORDS_START_CHARS + "])(?:" + String.join("|", KEYWORDS) + ")\\b")
+        .or(STRING.css, "\"[^\"\\\\]*(\\\\.[^\"\\\\]*)*\"")
+        .or(CHAR.css, "'(?:[^']|\\\\(?:'|u\\w{4}))'") // char
+        .or(NULL.css, "\\bnull\\b") 
+        .or(BOOLEAN.css, "\\btrue|false\\b") 
+        .or("annotation", "@[\\w]+")
+        .or("class-ident", "\\b[A-Z][\\w_$]*\\b")
```

Comment:
The way I see it, the enum can only help two languages sharing a common construct use the same style (eg every language uses the `string` css class for their strings). Language specific constructs like Java annotations, XPath axes or the like cannot be shared between languages so I'd say it's useless to make an enum key for them. Wdyt?

Reply:
I was personally thinking not so much about sharing between syntax, but for making sure:

1. that whatever we type here exists on the css
2. we can easily check which grammars use a specific rule (ie: to remove it)
3. allow to easily rename rules

If we care about reducing the number of rules, we may simply change these to more abstract / semantic names we can use at our own discretion at the grammar level. For instance we may simply have a "literal" rule, and decide to apply it depending on the language to numbers, true / false, null, etc. instead of 3 different rules as we currently have (also, all 3 look the same, so no difference there). Such approach may be a little more limited, but would allow to keep number of classes under control to as few as we may want (comment, literal value, string literal, keyword, annotation and normal text?)

Also, annotations are part of Apex too (see https://github.com/pmd/pmd/blob/master/pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ASTAnnotation.java). They are even used by PMD to enable violation suppression, like in Java.

Reply:
You make a good case. I'll update the enum

I think a good compromise between flexibility and css size would be to allow each token to be tagged with several classes. That way, numbers, booleans, null and char literals could all be tagged with a `literal` class, thus defining a base for all of them, but if we want to specialise the rendering of eg null, we can still do it if also has a `null` class. I imagine the enum could then declare such constants as eg `NULL(".null.literal")`, and the classes would be split in the highlight grammar. 

> Also, annotations are part of Apex too

I don't know how I could have missed that, you're right